skorch/helper.py

```diff
@@ -131,6 +132,105 @@ def __ne__(self, other):
         return not self.__eq__(other)
 
 
+# This class must be an instance of Sequence and have an ndim
+# attribute because sklearn will test this.
+class SliceDataset(Sequence):
+    # pylint: disable=anomalous-backslash-in-string
+    """Helper class that wraps a torch dataset to make it work with
+    sklearn.
+
+    Sometimes, sklearn will touch the input data, e.g. when splitting
+    the data for a grid search. This will fail when the input data is
+    a torch dataset. To prevent this, use this wrapper class for your
+    dataset.
+
+    Note: This class will only return the X value by default (i.e. the
+    first value returned by indexing the original dataset). Sklearn,
+    and hence skorch, always require 2 values, X and y. Therefore, you
+    still need to provide the y data separately.
+
+    Note: This class behaves similarly to a PyTorch
+    :class:`~torch.utils.data.Subset` when it is indexed by a slice or
+    numpy array: It will return another ``SliceDataset`` that
+    references the subset instead of the actual values. Only when it
+    is indexed by an int does it return the actual values. The reason
+    for this is to avoid loading all data into memory when sklearn,
+    for instance, creates a train/validation split on the
+    dataset. Data will only be loaded in batches during the fit loop.
+
+    Examples
+    --------
+    >>> X = MyCustomDataset()
+    >>> search = GridSearchCV(net, params, ...)
+    >>> search.fit(X, y)  # raises error
+    >>> ds = SliceDataset(X)
+    >>> search.fit(ds, y)  # works
+
+    Parameters
+    ----------
+    dataset : torch.utils.data.Dataset
+      A valid torch dataset.
+
+    n : int (default=0)
+      Indicates which element of the dataset should be
+      returned. Typically, the dataset returns both X and y
+      values. SliceDataset can only return 1 value. If you want to
+      get X, choose n=0 (default), if you want y, choose n=1.
+
+    indices : list, np.ndarray, or None (default=None)
+      If you only want to return a subset of the dataset, indicate
+      which subset that is by passing this argument. Typically, this
+      can be left to be None, which returns all the data. See also
+      :class:`~torch.utils.data.Subset`.
+
+    """
+    def __init__(self, dataset, n=0, indices=None):
+        self.dataset = dataset
+        self.n = n
+        self.indices = indices
+
+        self.indices_ = (self.indices if self.indices is not None
+                         else np.arange(len(self.dataset)))
```

Comment:
Is the `indices=None` case tested?

Reply:
Yes, as it's the default. But I added a test for `indices != None`